### PR TITLE
feat(KRV-06): refresh register route

### PR DESCRIPTION
**Pull request by Sanarielsen**
Refresh register route with roles dinamically can be used to and two validations, these includes:
1. Verify if all params sent for you in register;
2. Verify if current email informed is current used in DB for other user;

### CODE REVIEW

File: src/Controller/RegistrationController.php [hunk 1]
```diff
@@ -26,40 +25,37 @@ public function __construct(EmailVerifier $emailVerifier)
...
             );
+            $user->setRoles($resultJson["roles"]);
+
+            $userArray = get_object_vars($user);
+            if ( count($userArray) <= 3 ) {
+                return $this->json(['status' => '400', 'message' => 'Erro ao prosseguir com esse cadastro. Preencha todos os campos para prosseguir cadastro'], 200, ['Content-Type'=>'application/json; charset=utf-8']);
+            }
+            $userSearched = $entityManager->getRepository(User::class)->findUserByEmail($user->getEmail());
+            if ( $userSearched > 0 ) {
+                return $this->json(['status' => '400', 'message' => 'Erro ao prosseguir com esse cadastro. O seguinte campo deve ser único: email'], 200, ['Content-Type'=>'application/json; charset=utf-8']);
```
Comment (LovlsGenesis) [inline src/Controller/RegistrationController.php]: Bad practice. You can't give the information that an email is already being used. It helps hacker to know that an email exist then they can try to brute force the password to hack the user.

File: src/Repository/UserRepository.php [hunk 2]
```diff
@@ -40,6 +40,17 @@ public function upgradePassword(PasswordAuthenticatedUserInterface $user, string
         $this->getEntityManager()->flush();
     }
 
+    public function findUserByEmail($email): array
+    {
+        return $this->createQueryBuilder('u')
+            ->andWhere('u.email = :val')
+            ->setParameter('val', $email)
+            ->orderBy('u.id', 'ASC')
+            ->setMaxResults(1)
```
Comment (LovlsGenesis) [inline src/Repository/UserRepository.php]: Your variable should be easy to understand. 
`u` doesn't mean anything.

File: src/Controller/RegistrationController.php [hunk 3]
```diff
@@ -26,40 +25,37 @@ public function __construct(EmailVerifier $emailVerifier)
...
+                    $resultJson["password"]
                 )
             );
+            $user->setRoles($resultJson["roles"]);
+
+            $userArray = get_object_vars($user);
+            if ( count($userArray) <= 3 ) {
+                return $this->json(['status' => '400', 'message' => 'Erro ao prosseguir com esse cadastro. Preencha todos os campos para prosseguir cadastro'], 200, ['Content-Type'=>'application/json; charset=utf-8']);
+            }
+            $userSearched = $entityManager->getRepository(User::class)->findUserByEmail($user->getEmail());
```
Comment (LovlsGenesis) [inline src/Controller/RegistrationController.php]: 'Preencha todos os campos para prosseguir cadastro' -> 'Preencha todos os campos para prosseguir **o** cadastro'

Comment (LovlsGenesis) [inline src/Controller/RegistrationController.php]: foundUser or userFound would have been better

File: src/Controller/RegistrationController.php [hunk 4]
```diff
@@ -26,40 +25,37 @@ public function __construct(EmailVerifier $emailVerifier)
...
+            if ( $userSearched > 0 ) {
```
Comment (LovlsGenesis) [inline src/Controller/RegistrationController.php]: Does it really work? You ain't returning a number from the `$userSearched = $entityManager->getRepository(User::class)->findUserByEmail($user->getEmail());`

Shouldn't you validate that this variable isn't empty ? 

EDIT: 

`//     * @return User[] Returns an array of User objects`

File: src/Controller/RegistrationController.php [hunk 5]
```diff
@@ -26,40 +25,37 @@ public function __construct(EmailVerifier $emailVerifier)
...
-                    ->to($user->getEmail())
-                    ->subject('Please Confirm your Email')
-                    ->htmlTemplate('registration/confirmation_email.html.twig')
-            );
-            // do anything else you need here, like send an email
-
-            return $this->redirectToRoute('_profiler_home');
+            return $this->json(['status' => '200', 'message' => 'Funcionário cadastrado com sucesso.'], 200, ['Content-Type'=>'application/json; charset=utf-8']);
+        } catch (Exception $e) {
+            throw $this->createAccessDeniedException('Cannot create new users with current credentials');
```
Comment (LovlsGenesis) [inline src/Controller/RegistrationController.php]: new user*

**Review by LovlsGenesis — CHANGES_REQUESTED** [review]

File: src/Controller/RegistrationController.php (hunk 1 above)
Comment (Sanarielsen) [reply src/Controller/RegistrationController.php]: I have a question. How can I send to my front end which error I have to register user? Or can return a list with possibles things cause that error?

Obs: That screen only can access for admin users.

File: src/Repository/UserRepository.php (hunk 2 above)
Comment (Sanarielsen) [reply src/Repository/UserRepository.php]: I fix that named "u" for "user"

File: src/Controller/RegistrationController.php (hunk 4 above)
Comment (Sanarielsen) [reply src/Controller/RegistrationController.php]: Yes, I validate with empty value and worked correctly

File: src/Controller/RegistrationController.php [hunk 6]
```diff
@@ -26,40 +25,37 @@ public function __construct(EmailVerifier $emailVerifier)
...
             $user->setPassword(
                 $userPasswordHasher->hashPassword(
                     $user,
-                    $form->get('plainPassword')->getData()
+                    $resultJson["password"]
                 )
             );
+            $user->setRoles($resultJson["roles"]);
+
+            $userArray = get_object_vars($user);
```
Comment (LovlsGenesis) [inline src/Controller/RegistrationController.php]: Rename this variable for userData or userParams

**Review by LovlsGenesis — CHANGES_REQUESTED** [review]

**Review by LovlsGenesis — APPROVED** [review]

**Pull request merged**